File: v2/dtplan_web/src/app/core/services/api/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Router } from '@angular/router';
import { catchError, map, tap } from 'rxjs/operators';
import { Observable, of, throwError } from 'rxjs';

export interface CadastroDTO {
  email: string;
  senha: string;
  //permissao: string;
  nome: string;
  cpf: string;
  tipoUsuario: string;
   crn: string;
   cref: string;
}

export interface LoginDTO {
  email: string;
  senha: string;
}

export interface TokenDTO {
  token: string;
  refreshToken: string;
  id: number;
  nome: string;
}

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private apiUrl = 'http://localhost:8081'; // URL do backend

  constructor(private http: HttpClient, private router: Router) {}

  login(email: string, senha: string): Observable<TokenDTO> {
    const loginData: LoginDTO = { email, senha };
    return this.http.post<TokenDTO>(`${this.apiUrl}/auth/login`, loginData).pipe(
      tap((response) => {
        localStorage.setItem('token', response.token);
        localStorage.setItem('usuario', JSON.stringify({ id: response.id, nome: response.nome }));
        localStorage.setItem('login', email);
      }),
      catchError(this.handleError)
    );
  }

  validarToken(token: string): Observable<boolean> {
    return this.http.post<{ isValid: boolean }>(`${this.apiUrl}/auth/token`, { token }).pipe(
      map((response) => response.isValid),
      catchError(this.handleError)
    );
  }

  logout(): void {
    localStorage.removeItem('token');
    localStorage.removeItem('usuario');
    this.router.navigate(['/login']);
  }

  isLoggedIn(): boolean {
    return !!localStorage.getItem('token');
  }

  getToken(): string {
    return localStorage.getItem('token') || '';
  }

  getUsuario(): { id: number; nome: string } | null {
    const usuario = localStorage.getItem('usuario');
    return usuario ? JSON.parse(usuario) : null;
  }

  private handleError(error: HttpErrorResponse) {
    console.error('Erro na requisição:', error);
    return throwError(() => new Error('Erro ao realizar a requisição. Tente novamente mais tarde.'));
  }
}

File: rascunho/dieta_e_treino/lib/services/perfil/auth_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';

class AuthService {
  // URL de login
  static const String loginUrl = 'http://10.0.2.2:8080/login';

  Future<String> login(String login, String senha) async {
    try {
      final response = await http.post(
        Uri.parse(loginUrl),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'login': login, 'senha': senha}),
      );

      if (response.statusCode == 200) {
        var data = jsonDecode(response.body);
        String token = data['token'];
        print(token);
        return token;
      } else {
        throw Exception('Erro de login: ${response.body}');
      }
    } catch (e) {
      throw Exception('Erro de conexão: $e');
    }
  }
}
